File: 61/solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def rotate_right(head, k)
    return nil if head.nil?
    n, counter_ptr = 0, head
    until counter_ptr.nil?
        n += 1
        counter_ptr = counter_ptr.next
    end
    
    k %= n
    fast = slow = head
    k.times { fast = fast.next unless fast.nil? }
    return head if fast.nil?
    until fast.next.nil?
        fast, slow = fast.next, slow.next
    end
    
    fast.next = head
    fast = slow.next
    slow.next = nil
    
    fast
end

File: 61/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        n, counterPtr = 0, head
        while counterPtr:
            n += 1
            counterPtr = counterPtr.next
            
        k %= n
        fast = slow = head
        for i in range(k):
            if fast:
                fast = fast.next
        if not fast:
            return head
        while fast.next:
            fast, slow = fast.next, slow.next
        
        fast.next = head
        fast = slow.next
        slow.next = None
        
        return fast
